tests/test_page_settings.py

```diff
@@ -0,0 +1,102 @@
+
+MISSING_ENTITIES = {
+    "foobar-bot": {
+        "connector": "GitHub",
+        "public_repositories": ["my_org/repo", "my_org/missing-repo"],
+    },
+    "boards": {
+        "connector": "Jira",
+        "membership_boards": ["Tmiss"],
+    },
+}
+
+
+def has_entity_id_imported_side_effect(*args, **kwargs):
+    if "miss" in kwargs.get("text", "") and kwargs.get("exact") is True:
+        mock_for_missing_entry = MagicMock()
+        # chain: .nth(0).is_visible() => False
+        mock_for_missing_entry.nth.return_value.is_visible.return_value = False
+        return mock_for_missing_entry
+    else:
+        mock_for_existing_entry = MagicMock()
+        # chain: .nth(0).is_visible() => True
+        mock_for_existing_entry.nth.return_value.is_visible.return_value = True
+        return mock_for_existing_entry
+
+
+def test_sync_repositories():
+    for integrations in [GH_INTEGRATION_MEMBERSHIP, GH_INTEGRATION]:
+        with patch("logilica_weekly_report.page_settings.expect") as mock_expect:
+            page_mock = MagicMock()
+            page_mock.get_by_text.side_effect = has_entity_id_imported_side_effect
+            mock_expect.return_value = page_mock
+            page = SettingsPage(page=page_mock)
+            page.sync_integrations(integrations=integrations)
+            page_mock.get_by_role.assert_any_call("heading", name="GitHub Settings")
+
+
+def test_sync_boards():
+    with patch("logilica_weekly_report.page_settings.expect") as mock_expect:
+        page_mock = MagicMock()
+        page_mock.get_by_text.side_effect = has_entity_id_imported_side_effect
+
+        mock_locator = MagicMock()
+        mock_locator.count.return_value = 1
+        page_mock.get_by_role.return_value.filter.return_value.get_by_role.return_value = (
+            mock_locator
+        )
+
+        mock_expect.return_value = page_mock
+        page = SettingsPage(page=page_mock)
+        page.sync_integrations(integrations=JIRA_INTEGRATION)
+        print(mock_locator.mock_calls)
+        page_mock.get_by_role.assert_any_call("heading", name="Jira Settings")
+
+
+def test_missing_entities():
+    with patch("logilica_weekly_report.page_settings.expect") as mock_expect:
+        page_mock = MagicMock()
+        page_mock.get_by_text.side_effect = has_entity_id_imported_side_effect
+
+        mock_locator = MagicMock()
+        mock_locator.count.return_value = 1
+        page_mock.get_by_role.return_value.filter.return_value.get_by_role.return_value = (
+            mock_locator
+        )
+
+        mock_expect.return_value = page_mock
+        page = SettingsPage(page=page_mock)
+
+        with raises(RuntimeError) as error:
+            page.sync_integrations(integrations=MISSING_ENTITIES)
+
+        assert (
+            sum(len(value_list) for value_list in error.value.args[0].values()) == 2
+        ), "There are two integrations with 2 failures that failed to sync"
```

Comment:
We only see the supplied text when the assertion _fails_, which would be when the number of integration failures is **_not_** two, right?  So, I think the supplied text should probably say _that_.  Ideally, the assertion should report what the expected value was and what the actual value was.

Also, should we be making an assertion about the user-visible text of the `RuntimeError` (i.e., to confirm that the `RuntimeError` instance that we received was the expected one)?

Reply:
We don't need to do that, if there are sync failures, it raises `RuntimeError(sync_failures)`

So `error.value.args[0]` returns `defaultdict[Tuple[str, str], list[Tuple[str, str]]]` and that's where we get `.values()`. Other RuntimeError instance would not have that.